plugins/none-ls/_mk-source-plugin.nix

```
@@ -0,0 +1,53 @@
+# mkSourcePlugin, returns a module
+sourceType: sourceName:
+{
+  config,
+  pkgs,
+  lib,
+  helpers,
+  ...
+}:
+let
+  inherit (import ./packages.nix pkgs) packaged unpackaged;
+
+  cfg = config.plugins.none-ls;
+  cfg' = config.plugins.none-ls.sources.${sourceType}.${sourceName};
+in
+{
+  options.plugins.none-ls.sources.${sourceType}.${sourceName} =
+    {
+      enable = lib.mkEnableOption "the ${sourceName} ${sourceType} source for none-ls";
+      withArgs = helpers.mkNullOrOption helpers.nixvimTypes.strLua ''
+        Raw Lua code passed as an argument to the source's `with` method.
+      '';
+    }
+    # Only declare a package option if a package is required
+    // lib.optionalAttrs (packaged ? ${sourceName} || lib.elem sourceName unpackaged) {
```

@traxys can you clarify why it is useful to have a separate `unpackaged` list?

If we made `packaged` into a nullable list, none of _this_ implementation (below) would have to change, but this line could be simplified:

```suggestion
    // lib.optionalAttrs (packaged ? ${sourceName}) {
```

If you just like the _list syntax_ and not needing to spam `null` repeatedly, we can use `genAttrs`:

```nix
packaged =
  {
    inherit (pkgs)
      actionlint
      alejandra
      # etc
      ;
  }
  # Unpackaged
  // genAttrs [
    "blade_formatter"
    "bsfmt"
    # etc
  ] (_: null);
```

(I agree the _package or null or boolean_ we had before was clunky, so I see the value in the separate `noPackage` list).

---

(I've pushed up a change implementing this, let me know if you'd rather I drop/revert it)

---

I don't have any issues with that, it was just simpler to write